### PR TITLE
feat: add investor count compliance module

### PR DESCRIPTION
**Pull request by daanporon**
## Summary

Implements InvestorCountComplianceModule to enforce regulatory investor limits for asset tokenization:

- **Global and per-country caps**: Configure limits like "max 100 total with 50 US, 30 EU limits"
- **Topic-based filtering**: Only count investors with specific compliance claims (e.g., KYC verified)  
- **Real-time tracking**: Automatically tracks investors through token lifecycle events (transfer, mint, burn)
- **Cross-token enforcement**: Support issuer-wide limits across multiple token contracts
- **Comprehensive testing**: Full test coverage including edge cases and mock utilities

## Key Features

- Flexible configuration supporting various regulatory scenarios
- Efficient storage with separate token-specific and global tracking
- Integration with SMART identity registry for compliance validation
- Support for private placements and security offering requirements
- Country-specific limits based on ISO country codes

## Summary by Sourcery

Add InvestorCountComplianceModule for regulatory enforcement of investor count limits with flexible configuration (global, country, topic-based), real-time lifecycle tracking, and public view methods, backed by a full suite of tests and a mock token implementation.

New Features:
- Introduce InvestorCountComplianceModule to enforce unique investor limits per token and issuer with global and country-specific caps
- Enable expression-based topic filtering to count only compliant investors (e.g., KYC, AML, collateral)
- Track investor lifecycle in real time via mint, transfer, and burn hooks
- Expose view functions to query total and per-country investor counts and investor status

Tests:
- Add comprehensive test suite covering parameter validation, transfer logic, country-specific and global caps, topic filters, lifecycle events, view functions, and edge cases
- Introduce MockSMARTToken to simulate balances for direct compliance module testing

## Summary by Sourcery

Introduce a new InvestorCountComplianceModule to enforce regulatory investor count limits across smart tokens, and update the subgraph, tooling, and tests to support indexing, decoding, and querying its parameters and lifecycle events.

New Features:
- Add InvestorCountComplianceModule smart contract to enforce global and country-specific investor caps with topic-based filtering and issuer-wide tracking
- Extend subgraph schema and fetch logic to store InvestorCountParams and related ExpressionNode trees
- Register and integrate the new module in type identifiers, genesis artifacts, Hardhat/Ignition ignition scripts, and code generation configuration

Tests:
- Add a comprehensive test suite for InvestorCountComplianceModule covering parameter validation, limit enforcement, transfer/mint/burn lifecycle, topic filter expressions, global tracking, and view functions

Chores:
- Introduce MockSMARTToken for simulating balances in compliance module tests
- Update artifacts-genesis, hardhat constants, codegen-types, and ignition modules to include the new investor count compliance module

### CODE REVIEW

File: kit/contracts/contracts/smart/modules/InvestorCountComplianceModule.sol [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+import { ExpressionNode } from "../interface/structs/ExpressionNode.sol";
+
+/// @title Investor Count Compliance Module
+/// @author SettleMint
+/// @notice Restricts the number of unique investors who can hold tokens based on jurisdictional limits
+/// @dev Tracks unique investors (addresses with balance > 0) and enforces both global and per-country limits.
+///      Uses ExpressionNode filtering to ensure only investors meeting specific compliance requirements
+///      are counted toward the limits. This module is essential for regulatory compliance in
+///      jurisdictions with investor count restrictions (e.g., private placements, security offerings).
+/// @custom:security-contact security@settlemint.com
```
[the remark this answers is not in the document]
Comment (daanporon) [reply kit/contracts/contracts/smart/modules/InvestorCountComplianceModule.sol:19]: we use onlyTokenOrCompliance(_token), so that only known contracts can call it
